src/compiler/checker.ts

```diff
@@ -23732,7 +23732,7 @@ namespace ts {
 
             function getNarrowedType(type: Type, candidate: Type, assumeTrue: boolean, isRelated: (source: Type, target: Type) => boolean) {
                 if (!assumeTrue) {
-                    return filterType(type, t => !isRelated(t, candidate));
+                    return filterType(type, t => !isRelated(t, candidate), /*checkConstraints*/ true);

```

Comment:
Do you anticipate this parameter being useful outside this one call? Why not just put the constraint-checking code in this one callback?

Reply:
> Do you anticipate this parameter being useful outside this one call?

OK, so, get this: I _removed_ this flag and made `filterType` _unconditionally_ break apart constraints. We use `filterType` all over - surely this would make something fail in a case where, eg, we're looking for nullable types and start returning arbitrary `T`s constrained to nullable types. Surely that would cause a test to fail. And yet, it did not. And this was spooky - is our coverage of the different ways we filter types really so poor? So I kept the flag, not wanting to accidentally break other usages of the `filterType` call that maybe did not expect the constraint decomposition when they were written, but are just poorly covered by tests.

> Why not just put the constraint-checking code in this one callback?

I guess I could! It'd arguably be more consistent then, especially since it'd cause the constraint breakdown to be recursively invoked on unions whose members are type variables.